File: src/types/ai.ts
export type AiAction = "Summarize" | "SolveProblem" | "Ask" | "GenerateChatTitle";
export type TerminalAgentExecutionTarget = "TerminalWindow" | "ChatWindow";
export type TerminalAgentPhase =
  | "Starting"
  | "Probing"
  | "Planning"
  | "AwaitingApproval"
  | "AwaitingPassword"
  | "RunningCommands"
  | "Done"
  | "Blocked"
  | "Cancelled"
  | "Failed";
export type TerminalAgentPlanPhase =
  | "Starting"
  | "Probing"
  | "Questioning"
  | "AwaitingAnswers"
  | "GeneratingOptions"
  | "AwaitingSelection"
  | "ReadyToExecute"
  | "Done"
  | "Blocked"
  | "Cancelled"
  | "Failed";
export type TerminalAgentRisk = "read_only" | "requires_confirmation";
export type TerminalAgentEventKind =
  | "command_started"
  | "stdout"
  | "stderr"
  | "command_finished";

export type AiTokenizerType =
  | "Estimate"
  | "Cl100kBase"
  | "O200kBase"
  | "P50kBase"
  | "R50kBase";

export type AiTokenLimitUnit = "Thousands" | "Millions";
export type AiTokenWarningLevel = "None" | "Yellow" | "Red";
export type AiChatRole = "User" | "Assistant";

export interface AiProfile {
  id: string;
  name: string;
  apiUrl: string;
  model: string;
  apiKey: string;
  maxSelectionChars: number;
  tokenizerType: AiTokenizerType;
  tokenLimitAmount?: number;
  tokenLimitUnit: AiTokenLimitUnit;
  tokenWarningYellowPercent: number;
  tokenWarningRedPercent: number;
  tokenResetPeriodDays: number;
  tokenResetAnchorDate?: string;
  tokenUsageCycleStartDate?: string;
  usedPromptTokens: number;
  usedCompletionTokens: number;
  usedTotalTokens: number;
}

export interface AiTokenUsage {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
}

export interface AiTokenUsageSnapshot {
  usedPromptTokens: number;
  usedCompletionTokens: number;
  usedTotalTokens: number;
  maxTokens: number;
  remainingTokens?: number;
  cycleStartDate: string;
  nextResetDate: string;
  warningLevel: AiTokenWarningLevel;
  unlimited: boolean;
}

export interface AiExecutionResult {
  content: string;
  usage?: AiTokenUsage;
  usageSnapshot?: AiTokenUsageSnapshot;
  activeProfileId?: string;
  activeProfileName?: string;
}

export interface TerminalAgentRequest {
  sessionId: string;
  profileId: string;
  userPrompt: string;
  connectionDisplayName?: string;
  acceptedPlanContext?: string;
  executionTarget: TerminalAgentExecutionTarget;
  showDebugMessages: boolean;
  showRuntimeMessages: boolean;
  askConfirmationBeforeEveryCommand: boolean;
  autoApproveRootCommands: boolean;
}

export interface TerminalAgentPlanRequest {
  sessionId: string;
  profileId: string;
  userPrompt: string;
  connectionDisplayName?: string;
}

export interface TerminalAgentPlanStartResponse {
  runId: string;
  initialState: TerminalAgentPlanRunState;
}

export interface TerminalAgentStartResponse {
  runId: string;
}

export interface TerminalAgentPlanExecutionResponse {
  runId: string;
  request: TerminalAgentRequest;
}

export interface TerminalAgentPlannedCommand {
  command: string;
  purpose: string;
  risk: TerminalAgentRisk;
}

export interface TerminalAgentApproval {
  runId: string;
  sessionId: string;
  executionTarget: TerminalAgentExecutionTarget;
  summary: string;
  userMessage: string;
  commands: TerminalAgentPlannedCommand[];
}

export interface TerminalAgentPasswordRequest {
  runId: string;
  sessionId: string;
  executionTarget: TerminalAgentExecutionTarget;
  summary: string;
  userMessage: string;
  command: string;
}

export interface TerminalAgentProbeSnapshot {
  osRelease: string;
  kernel: string;
  architecture: string;
  shell: string;
  currentUser: string;
  uid: string;
  gid: string;
  groups: string[];
  homeDir: string;
  currentDir: string;
  availableDiskKb?: number;
  availableDiskPath: string;
  packageManagers: string[];
  serviceManagers: string[];
  alreadyRoot: boolean;
  sudoAvailable: boolean;
  passwordlessSudo: boolean;
  sudoNonInteractive: boolean;
  sudoNListSummary: string;
  rootEscalationMode: string;
}

export interface TerminalAgentPlanQuestion {
  id: string;
  question: string;
}

export interface TerminalAgentPlanOption {
  id: string;
  title: string;
  summary: string;
  feasibility: string;
  risks: string[];
  prerequisites: string[];
  steps: string[];
  alternatives: string[];
}

export interface TerminalAgentPlanQuestionsEvent {
  runId: string;
  sessionId: string;
  questions: TerminalAgentPlanQuestion[];
}

export interface TerminalAgentPlanOptionsEvent {
  runId: string;
  sessionId: string;
  options: TerminalAgentPlanOption[];
  acceptedOptionId?: string;
}

export interface TerminalAgentPlanRunState {
  runId: string;
  sessionId: string;
  phase: TerminalAgentPlanPhase;
  summary: string;
  userMessage?: string;
  probeSummary?: string;
  questions?: TerminalAgentPlanQuestion[];
  options?: TerminalAgentPlanOption[];
  acceptedOptionId?: string;
  executionStartedRunId?: string;
}

export interface TerminalAgentCommandResult {
  command: string;
  purpose: string;
  risk: TerminalAgentRisk;
  exitStatus?: number;
  exitSignal?: string;
  stdoutTail: string;
  stderrTail: string;
  stdoutTruncated: boolean;
  stderrTruncated: boolean;
  cancelled: boolean;
  timedOut: boolean;
}

export interface TerminalAgentRunState {
  runId: string;
  sessionId: string;
  executionTarget: TerminalAgentExecutionTarget;
  phase: TerminalAgentPhase;
  summary: string;
  userMessage?: string;
  pendingApproval?: TerminalAgentApproval;
  pendingPasswordRequest?: TerminalAgentPasswordRequest;
  currentCommand?: string;
  turn: number;
}

export interface TerminalAgentEvent {
  runId: string;
  sessionId: string;
  executionTarget: TerminalAgentExecutionTarget;
  kind: TerminalAgentEventKind;
  command?: string;
  purpose?: string;
  chunk?: string;
  result?: TerminalAgentCommandResult;
}

export interface AiRequestPayload {
  action: AiAction;
  profileId: string;
  selectedText: string;
  connectionDisplayName?: string;
  responseLanguageCode?: string;
  userPrompt?: string;
  conversationContext?: string;
}

export interface SavedAiChatMessage {
  role: AiChatRole;
  content: string;
  createdAt: number;
  aiProfileId?: string;
  aiProfileName?: string;
}

export interface SavedAiChat {
  id: string;
  title: string;
  createdAt: number;
  updatedAt: number;
  selectedText: string;
  connectionDisplayName?: string;
  responseLanguageCode?: string;
  activeAiProfileId?: string;
  activeAiProfileName?: string;
  messages: SavedAiChatMessage[];
}

export function createEmptyAiProfile(): AiProfile {
  return {
    id: crypto.randomUUID(),
    name: "",
    apiUrl: "",
    model: "",
    apiKey: "",
    maxSelectionChars: 1_000_000,
    tokenizerType: "Estimate",
    tokenLimitUnit: "Thousands",
    tokenWarningYellowPercent: 75,
    tokenWarningRedPercent: 90,
    tokenResetPeriodDays: 30,
    usedPromptTokens: 0,
    usedCompletionTokens: 0,
    usedTotalTokens: 0,
  };
}

File: src-tauri/src/model/ai.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum AiAction {
    Summarize,
    SolveProblem,
    Ask,
    GenerateChatTitle,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub enum AiTokenizerType {
    #[default]
    Estimate,
    Cl100kBase,
    O200kBase,
    P50kBase,
    R50kBase,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub enum AiTokenLimitUnit {
    #[default]
    Thousands,
    Millions,
}

impl AiTokenLimitUnit {
    pub fn multiplier(&self) -> u64 {
        match self {
            Self::Thousands => 1_000,
            Self::Millions => 1_000_000,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum AiTokenWarningLevel {
    None,
    Yellow,
    Red,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiProfile {
    pub id: String,
    pub name: String,
    pub api_url: String,
    pub model: String,
    #[serde(default)]
    pub api_key: String,
    #[serde(default = "default_max_selection_chars")]
    pub max_selection_chars: usize,
    #[serde(default)]
    pub tokenizer_type: AiTokenizerType,
    pub token_limit_amount: Option<u64>,
    #[serde(default)]
    pub token_limit_unit: AiTokenLimitUnit,
    #[serde(default = "default_yellow_percent")]
    pub token_warning_yellow_percent: u8,
    #[serde(default = "default_red_percent")]
    pub token_warning_red_percent: u8,
    #[serde(default = "default_reset_days")]
    pub token_reset_period_days: u16,
    pub token_reset_anchor_date: Option<String>,
    pub token_usage_cycle_start_date: Option<String>,
    #[serde(default)]
    pub used_prompt_tokens: u64,
    #[serde(default)]
    pub used_completion_tokens: u64,
    #[serde(default)]
    pub used_total_tokens: u64,
}

impl Default for AiProfile {
    fn default() -> Self {
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name: String::new(),
            api_url: String::new(),
            model: String::new(),
            api_key: String::new(),
            max_selection_chars: default_max_selection_chars(),
            tokenizer_type: AiTokenizerType::Estimate,
            token_limit_amount: None,
            token_limit_unit: AiTokenLimitUnit::Thousands,
            token_warning_yellow_percent: default_yellow_percent(),
            token_warning_red_percent: default_red_percent(),
            token_reset_period_days: default_reset_days(),
            token_reset_anchor_date: None,
            token_usage_cycle_start_date: None,
            used_prompt_tokens: 0,
            used_completion_tokens: 0,
            used_total_tokens: 0,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiTokenUsage {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub total_tokens: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiTokenUsageSnapshot {
    pub used_prompt_tokens: u64,
    pub used_completion_tokens: u64,
    pub used_total_tokens: u64,
    pub max_tokens: u64,
    pub remaining_tokens: Option<u64>,
    pub cycle_start_date: String,
    pub next_reset_date: String,
    pub warning_level: AiTokenWarningLevel,
    pub unlimited: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiExecutionResult {
    pub content: String,
    pub usage: Option<AiTokenUsage>,
    pub usage_snapshot: Option<AiTokenUsageSnapshot>,
    pub active_profile_id: Option<String>,
    pub active_profile_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiRequestPayload {
    pub action: AiAction,
    pub profile_id: String,
    pub selected_text: String,
    pub connection_display_name: Option<String>,
    pub response_language_code: Option<String>,
    pub user_prompt: Option<String>,
    pub conversation_context: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SavedAiChat {
    pub id: String,
    pub title: String,
    pub created_at: i64,
    pub updated_at: i64,
    pub selected_text: String,
    pub connection_display_name: Option<String>,
    pub response_language_code: Option<String>,
    pub active_ai_profile_id: Option<String>,
    pub active_ai_profile_name: Option<String>,
    #[serde(default)]
    pub messages: Vec<SavedAiChatMessage>,
}

impl Default for SavedAiChat {
    fn default() -> Self {
        let now = chrono::Utc::now().timestamp_millis();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            title: "AI Chat".into(),
            created_at: now,
            updated_at: now,
            selected_text: String::new(),
            connection_display_name: None,
            response_language_code: Some("en".into()),
            active_ai_profile_id: None,
            active_ai_profile_name: None,
            messages: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum AiChatRole {
    User,
    Assistant,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SavedAiChatMessage {
    pub role: AiChatRole,
    pub content: String,
    pub created_at: i64,
    pub ai_profile_id: Option<String>,
    pub ai_profile_name: Option<String>,
}

fn default_max_selection_chars() -> usize {
    1_000_000
}

fn default_yellow_percent() -> u8 {
    75
}

fn default_red_percent() -> u8 {
    90
}

fn default_reset_days() -> u16 {
    30
}
